### PR TITLE
Hardening Password Reset Process

### PR DESCRIPTION
**Pull request by ryanrath**
## Description
Harden the password reset process.

**NOTE:** When testing this change it is important to look at emails generated during user creation, triggering a password reset via the internal dashboard, and login -> 'Forgot password'. 

## Motivation and Context
MD5 can be brute forced fairly easily in the time before the link is expired. We now use SHA256 to calculate the hashes.

## Tests performed
Manual Testing via docker

## Types of changes

- [X] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [X] My code follows the code style of this project as found in the **CONTRIBUTING** document.
- [ ] I have added tests to cover my changes.
- [X] All new and existing tests passed.

### CODE REVIEW

**Comment by ryanrath** [comment]
@jpwhite4 Agreed, I've just pushed a commit based on your email. I've completed manual tests for: 
  - Happy Path
  - Cryptographic Splicing ( Attempting to change the data within the token to reset a different users password ) fails.
  - That expiration is enforced

These were done both in a fresh install & upgrade. I do still need to test XSEDE and OpenXDMoD w/ SUPREMM.

But if you could take a look when you have a second that would be great.

File: classes/Models/Services/Users.php [hunk 1]
```diff
@@ -418,4 +418,89 @@ public static function isPrincipalInvestigator(XDUser $user)
...
+        if ($now >= $expiration) {
+            $expirationDate = date('Y-m-d H:i:s', $expiration);
+            $log->debug("RID Token expired for: $userId | expired: $expirationDate");
+            return $results;
+        }
+
+        try {
+            $user = XDUser::getUserByID($userId);
+            $expected = explode('|', Users::generateRID($user, $expiration))[2];
+            $valid = \Symfony\Polyfill\Php56\Php56::hash_equals($expected, $hash);
```
Comment (jpwhite4) [inline classes/Models/Services/Users.php]: change this to 
```
$valid = hash_equals($expected, $hash);
```

File: composer.lock [hunk 2]
```diff
@@ -1,10 +1,10 @@
 {
     "_readme": [
         "This file locks the dependencies of your project to a known state",
-        "Read more about it at https://getcomposer.org/doc/01-basic-usage.md#composer-lock-the-lock-file",
```
Comment (jpwhite4) [inline composer.lock]: We need to not change all these dependencies this late in the development cycle.

File: html/controllers/user_auth/update_pass.php [hunk 3]
```diff
@@ -16,7 +18,7 @@
 	
 	// -----------------------------
 	
-   $validationCheck = XDUser::validateRID($_POST['rid']);
```
Comment (jpwhite4) [inline html/controllers/user_auth/update_pass.php]: Why did validateRID get added to Users? Also why is it still in XDUser?

Comment (ryanrath) [reply html/controllers/user_auth/update_pass.php]: - Because I felt that's where it belonged.
- Because I forgot to remove it.

File: classes/XDUser.php [hunk 4]
```diff
@@ -2989,4 +2957,88 @@ public function setOrganizationID($organizationID)
...
+
+        if ($now >= $expiration) {
+            $expirationDate = date('Y-m-d H:i:s', $expiration);
+            $log->debug("RID Token expired for: $userId | expired: $expirationDate");
+            return $results;
+        }
+
+        try {
+            $user = XDUser::getUserByID($userId);
+            $expected = explode('|', $user->generateRID($expiration))[2];
```
Comment (jpwhite4) [inline classes/XDUser.php]: getUserByID could return null -> you need to check for null return value

File: classes/XDUser.php [hunk 5]
```diff
@@ -2989,4 +2957,88 @@ public function setOrganizationID($organizationID)
...
+
+    /**
+     * Generate an RID value from the provided $user. This value is intended to be used in
+     * authenticating / authorizing a password reset. If an $expiration value is provided, that will
+     * be used instead of generating one via the 'email_token_expiration' portal settings value.
+     *
+     * @param int|null    $expiration the date after which this rid is considered invalid.
+     * @return string in the form "userId|expiration|hash"
+     * @throws Exception If there are any missing configuration properties that this function relies
+     * on. These include: email_token_expiration, email_token_expiration_format, and
```
Comment (jpwhite4) [inline classes/XDUser.php]: thisd documentation needs to be updated

File: classes/XDUser.php (hunk 4 above)
Comment (ryanrath) [reply classes/XDUser.php]: Inside the try?

File: html/controllers/user_auth/update_pass.php [hunk 6]
```diff
@@ -1,5 +1,7 @@
 <?php
 
+use Models\Services\Users;
```
Comment (jpwhite4) [inline html/controllers/user_auth/update_pass.php]: remove this change

File: html/password_reset.php [hunk 7]
```diff
@@ -7,14 +7,12 @@
    */
 
 	require_once dirname(__FILE__).'/../configuration/linker.php';
-
```
Comment (jpwhite4) [inline html/password_reset.php]: remove this change

File: html/controllers/user_admin/pass_reset.php [hunk 8]
```diff
@@ -1,6 +1,7 @@
 <?php
 
 use CCR\MailWrapper;
+use Models\Services\Users;
```
Comment (jpwhite4) [inline html/controllers/user_admin/pass_reset.php]: remove this change

File: classes/XDUser.php (hunk 4 above)
Comment (jpwhite4) [reply classes/XDUser.php]: the try will not catch the fatal error of dereferencing the null value. How did you test this case?

File: classes/XDUser.php [hunk 9]
```diff
@@ -2989,4 +2957,89 @@ public function setOrganizationID($organizationID)
...
+    public function generateRID($expiration = null)
+    {
+        $expiresIn = \xd_utilities\getConfiguration('general', 'email_token_expiration');
+        $appSecret = \xd_utilities\getConfiguration('general', 'application_secret');
+
+        $userId = $this->_id;
+        $userSecret = $this->_password;
+
+        if ($expiration === null) {
+            $expiration = time() + $expiresIn;
```
Comment (chakrabortyr) [inline classes/XDUser.php]: What happens when both expiration and the config setting are null? Should this have some default value?

Comment (ryanrath) [reply classes/XDUser.php]: So, what happens when somebody goes in and manually removes these settings?

File: html/controllers/user_auth/update_pass.php [hunk 10]
```diff
@@ -1,5 +1,6 @@
 <?php
 
+
```
Comment (chakrabortyr) [inline html/controllers/user_auth/update_pass.php]: Let's remove this.

File: html/password_reset.php [hunk 11]
```diff
@@ -1,174 +1,208 @@
 <!DOCTYPE HTML>
```
Comment (jpwhite4) [inline html/password_reset.php]: Please revert these change

Comment (ryanrath) [reply html/password_reset.php]: The reformatting? Or the validation checking?

File: open_xdmod/modules/xdmod/component_tests/lib/RIDTest.php [hunk 12]
```diff
@@ -0,0 +1,110 @@
...
+        );
+    }
+
+    /**
+     * @dataProvider provideGenerateRID
+     *
+     * @param array $options
+     */
+    public function testValidateRID(array $options) {
+
```
Comment (jpwhite4) [inline open_xdmod/modules/xdmod/component_tests/lib/RIDTest.php]: Please double check this function.

File: classes/XDUser.php [hunk 13]
```diff
@@ -2989,4 +2957,89 @@ public function setOrganizationID($organizationID)
...
+
+        try {
+            $user = XDUser::getUserByID($userId);
+            if ($user === null) {
+                return $results;
+            }
+            $expected = explode('|', $user->generateRID($expiration))[2];
+            $valid = hash_equals($expected, $hash);
+
+            $results['status'] = $valid ? 'success' : INVALID;
```
Comment (jpwhite4) [inline classes/XDUser.php]: update_pass.php checks for $validationCheck['status'] == VALID

File: classes/XDUser.php (hunk 9 above)
Comment (ryanrath) [reply classes/XDUser.php]: resolving per our conversation.
